HLTrigger/Timer/src/FastTimerService.cc

```diff
@@ -38,1425 +39,1365 @@
+{
+  auto const& stream = streams_[sid];
+  unsigned int pid = callgraph_.processId(process);
+  auto const& desc = callgraph_.processDescription(pid);
+  for (unsigned int id: boost::irange(0ul, desc.paths_.size()))
+    if (desc.paths_[id].name_ == path)
+      return ms(stream.processes[pid].paths[id].total.time_real);
+  for (unsigned int id: boost::irange(0ul, desc.endPaths_.size()))
+    if (desc.paths_[id].name_ == path)
+      return ms(stream.processes[pid].endpaths[id].total.time_real);
+
+  // FIXME issue a LogWarning, raise an exception, or return NaN
+  return 0.;
 }
 
-void
-FastTimerService::postStreamEndLumi(edm::StreamContext const & sc) {
-  unsigned int sid = sc.streamID().value();
-  auto & stream = m_stream[sid];
-
-  if (m_enable_dqm) {
-    DQMStore * store = edm::Service<DQMStore>().operator->();
-    assert(store);
-    store->mergeAndResetMEsLuminositySummaryCache(sc.eventID().run(),sc.eventID().luminosityBlock(),sid, m_module_id);
-  }
+double
+FastTimerService::queryHighlightTime(edm::StreamID sid, std::string const& label) const
+{
+  auto const& stream = streams_[sid];
+  for (unsigned int group: boost::irange(0ul, highlight_modules_.size()))
+    if (highlight_modules_[group].label == label)
+      return ms(stream.highlight[group].time_real);
 
-  stream.timer_last_transition = FastTimer::Clock::now();
+  // FIXME issue a LogWarning, raise an exception, or return NaN
+  return 0.;
 }
 
+
 void
-FastTimerService::postGlobalEndRun(edm::GlobalContext const & gc)
+FastTimerService::ignoredSignal(std::string signal) const
 {
-  if (m_enable_timing_summary) {
-    unsigned int pid = processID(gc.processContext());
-    unsigned int rid = gc.runIndex();
-    unsigned int run = gc.luminosityBlockID().run();
-    const std::string label = (boost::format("run %d") % run).str();
-
-    printProcessSummary(m_run_summary[rid], m_run_summary_perprocess[rid][pid], label, m_process[pid].name);
+  LogDebug("FastTimerService") << "The FastTimerService received is currently not monitoring the signal \"" << signal << "\".\n";
+}
 
-    if (pid+1 == m_process.size())
-      printSummary(m_run_summary[rid], label);
-  }
+void
+FastTimerService::unsupportedSignal(std::string signal) const
+{
+  // warn about each signal only once per job
+  if (unsupported_signals_.insert(signal).second)
+    edm::LogWarning("FastTimerService") << "The FastTimerService received the unsupported signal \"" << signal << "\".\n"
+      << "Please report how to reproduce the issue to cms-hlt@cern.ch .";
 }
 
 void
-FastTimerService::postEndJob()
+FastTimerService::preGlobalBeginRun(edm::GlobalContext const& gc)
 {
-  if (m_enable_timing_summary) {
-    const std::string label = "the whole job";
-    for (unsigned int pid = 0; pid < m_process.size(); ++pid)
-      printProcessSummary(m_job_summary, m_job_summary_perprocess[pid], label, m_process[pid].name);
+  ignoredSignal(__func__);
```

Comment:
Sorry, `ignoredSignal()` and `unsupportedSignal()` are more of a work-in-progress thing.

Eventually, I would like to assign all the time spent in the job to "something" - be it a run or lumi transition, and event, the framework itself, or account it as idle or blocked threads (though I don't really know how to do the last part).

For the moment, `unsupportSignal` are the signals I don't know what they do, while `ignoredSignal` are the signals I think I know what they do, but I am not handling yet.